File: Java/CodesInterviews/src/main/java/org/example/arrays/BestTimeToBuy.java
package org.example.arrays;

import java.util.Arrays;

public class BestTimeToBuy {

    public int maxProfit(int[] prices) {
        if (prices.length == 1) return 0;

        int profit = 0;
        int left = 0;
        for (int right = 1; right < prices.length; right++) {
            if (prices[right] > prices[left]) {
                profit = Math.max(profit, prices[right] - prices[left]);
            } else {
                left = right;
            }
        }
        return profit;
    } // TC: O(n), SC: O(1)

    public static void main(String[] args) {
        int[] prices = { 7, 1, 5, 3, 6, 4 };
        BestTimeToBuy b = new BestTimeToBuy();
        System.out.println(Arrays.toString(new int[]{b.maxProfit(prices)}));
    }
}

File: Kotlin/CodesInterviews/src/main/kotlin/arrays/BestTimeToBuy.kt
package arrays

class BestTimeToBuy {

    fun maxProfit(prices: IntArray): Int {
        if (prices.count() == 1) return 0

        var profit = 0
        var left = 0

        for (right in 1..<prices.count()) {
            if (prices[right] > prices[left]) {
                profit = Math.max(profit, prices[right]-prices[left])
            } else {
                left = right
            }
        }
        return profit
    }
}

fun main() {
    val b = BestTimeToBuy()
    val prices = intArrayOf(7, 1, 5, 3, 6, 4)
    println(b.maxProfit(prices))
}